app/src/json/json_parser.c

```diff
@@ -0,0 +1,255 @@
+/*******************************************************************************
+ *  (c) 2018 - 2024 Zondax AG
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ********************************************************************************/
+#include "json_parser.h"
+
+#include "../common/parser_common.h"
+#include "jsmn/jsmn.h"
+
+#define EQUALS(_P, _Q, _LEN) (MEMCMP((const void *)PIC(_P), (const void *)PIC(_Q), (_LEN)) == 0)
+
+/**
+ * Create JSON parser over an array of tokens
+ */
+extern void jsmn_init(jsmn_parser *parser);
+
+/**
+ * Run JSON parser. It parses a JSON data string into and array of tokens, each
+ * describing
+ * a single JSON object.
+ */
+extern int jsmn_parse(jsmn_parser *parser, const char *js, const size_t len, jsmntok_t *tokens,
+                      const unsigned int num_tokens);
+
+parser_error_t array_get_element_count(const parsed_json_t *json, uint16_t array_token_index, uint16_t *number_elements) {
+    *number_elements = 0;
+    if (array_token_index > json->numberOfTokens) {
+        return parser_no_data;
+    }
+
+    jsmntok_t array_token = json->tokens[array_token_index];
+    uint16_t token_index = array_token_index;
+    uint16_t prev_element_end = array_token.start;
+    while (true) {
+        token_index++;
+        if (token_index >= json->numberOfTokens) {
+            break;
+        }
+        jsmntok_t current_token = json->tokens[token_index];
+        if (current_token.start > array_token.end) {
+            break;
+        }
+        if (current_token.start <= prev_element_end) {
+            continue;
+        }
+        prev_element_end = current_token.end;
+        (*number_elements)++;
+    }
+
+    return parser_ok;
+}
+
+parser_error_t array_get_nth_element(const parsed_json_t *json, uint16_t array_token_index, uint16_t element_index,
+                                     uint16_t *token_index) {
+    if (array_token_index > json->numberOfTokens) {
+        return parser_no_data;
+    }
+
+    jsmntok_t array_token = json->tokens[array_token_index];
+    *token_index = array_token_index;
+
+    uint16_t element_count = 0;
+    uint16_t prev_element_end = array_token.start;
+    while (true) {
```

Comment:
why not using the token_index condition directly here? this works anyways just saying :-)

Reply:
while(*token_index < json->numberOfTokens)